src/libraries/System.Private.CoreLib/src/System/Reflection/NullabilityInfoContext.cs

```diff
@@ -385,12 +385,14 @@ private NullabilityInfo GetNullabilityInfo(MemberInfo memberInfo, Type type, ILi
 
                 for (int i = 0, offset = 0; i < genericArguments.Length; i++)
                 {
-                    if (!genericArguments[i].IsValueType)
+                    Type t = Nullable.GetUnderlyingType(genericArguments[i]) ?? genericArguments[i];
+
+                    if (!t.IsValueType || t.IsGenericType)
                     {
                         offset++;
                     }
 
-                    genericArgumentsState[i] = GetNullabilityInfo(memberInfo, genericArguments[i], customAttributes, offset);
+                    genericArgumentsState[i] = GetNullabilityInfo(memberInfo, genericArguments[i], customAttributes, index + offset);
                 }
```

Comment:
I'm confused by the logic.  Let's say I have:
```C#
Foo<Nullable<T1>, Nullable<T2>>
```
At i == 0, offset == 0, we'll increment offset (because T1 is generic), so then on L395 we'll store the info for the `Nullable<T1>` into genericArgumentsState[0], and we'll call GetNullabilityInfo with 0+1, which seems like it'll store the information for `T1` into genericArgumentsState[1].  Then on the next iteration, i == 1, and we'll end up overwriting genericArgumentState[1] with the info for `Nullable<T2>`.

Have I misread the code?

Reply:
> we'll store the info for the Nullable<T1> into genericArgumentsState[0], and we'll call GetNullabilityInfo with 0+1, which seems like it'll store the information for T1 into genericArgumentsState[1]

We are saving into `genericArgumentsState[i]` not `genericArgumentsState[index+offset]`, therefore no mixup with  `genericArguments` index

`index+offset` is only used for getting the index of the current `genericArgument` from `NullableAttribute` byte array

For your example `Foo<Nullable<T1>, Nullable<T2>>` the `NullableAttribute` would have only one value for nullability `Foo`, no bytes for type arguments because they are value type 

if it was `Foo<Nullable<int>, Foo1<string?>>` for example the  `NullableAttribute` byte array would be [1,1,2] where first byte is for `Foo<Nullable<int>, Foo1<string?>>`, second byte is for `Foo1<string?>` and last byte for <string?>. There is no byte for `Nullable<int>` parameter which nullability will not be evaluated from the `NullableAttribute`  byte array

Reply:
I see.

The C# compiler necessarily needs to do a similar parsing of the metadata. Does this code mirror what it does? Might there be other corner cases it's handling and we're missing?

Reply:
> The C# compiler necessarily needs to do a similar parsing of the metadata. Does this code mirror what it does?

This code is based on how the `NullableAttribute` byte array populated by the compiler with unit tests

> Might there be other corner cases it's handling and we're missing?
 
This and the [previous PR](https://github.com/dotnet/runtime/pull/58036) is fixing those corner cases. Reviewed the doc how the compiler populates `NullableAttribute` byte array: https://github.com/dotnet/roslyn/blob/main/docs/features/nullable-metadata.md#nullableattribute looks all scenarios in it covered/tested:

> The byte[] is constructed as follows:
> 
> - Reference type: the nullability (0, 1, or 2), followed by the representation of the type arguments in order including containing types
> - Nullable value type: the representation of the type argument only
> - Non-generic value type: skipped
> - Generic value type: 0, followed by the representation of the type arguments in order including containing types
> - Array: the nullability (0, 1, or 2), followed by the representation of the element type
> - Tuple: the representation of the underlying constructed type
> - Type parameter reference: the nullability (0, 1, or 2, with 0 for unconstrained type parameter)
>
> Note that non-generic value types are represented by an empty byte[]. However, generic value types and type parameters constrained to value types have an explicit 0 in the byte[] for nullability. The reason generic types and type parameters are represented with an explicit byte is to simplify metadata import. Specifically, this avoids the need to calculate whether a type parameter is constrained to a value type when decoding nullability metadata, since the constraints may include a (valid) cyclic reference to the type parameter.